### ISSUE TITLE
Implement new `QUERY` method to pass arguments in request body for read request

### ISSUE THREAD

**Issue by telewak** [PostgREST/postgrest#2125]
Thank you for this wonderful and magical api...  i would like to send alot of arguments while doing a GET request.
Does PostgREST as an API have a limit on the url length ? Is there possibility to POST a json object to the GET endpoint?

**Comment by wolfgangwalther** [comment]
See https://github.com/PostgREST/postgrest-docs/issues/417.

`POST` to `GET` is not implemented. Will keep this open to track the `QUERY` proposal by @steve-chavez in the thread above.

**Comment by steve-chavez** [comment]
Even the new QUERY http method won't help with cases where `DELETE` has a big `in` filter. PATCH would suffer from this too.

Perhaps we should have a POST with a special body + content type(similar to https://github.com/PostgREST/postgrest/issues/465) for these cases?

Come to think of it, QUERY would help because it will contain the DELETE/UPDATE inside.

**Comment by steve-chavez** [comment]
> Come to think of it, QUERY would help because it will contain the DELETE/UPDATE inside.
> I don't understand what you're saying here.

The example [here](https://www.ietf.org/archive/id/draft-ietf-httpbis-safe-method-w-body-02.html#section-4.2):

> QUERY /contacts HTTP/1.1
> Host: example.org
> Content-Type: example/query
> Accept: text/csv

> select surname, givenname, email limit 10

Hints that we could have a `delete where in (1,2,3,4,...)` or `update set name = 'xxx', status = 'yy' where in (1,2,3,4,..)` in the body? Maybe I'm wrong..

**Comment by steve-chavez** [comment]
> We'd use that for QUERY and DELETE by default -

You mean we would do an HTTP DELETE with a body? I thought that was frowned upon according to [this](https://stackoverflow.com/a/60357432/4692662).

**Comment by steve-chavez** [comment]
> Allow sending a POST to the root endpoint. This request would contain basically the whole request, including method/action, filters, headers, body etc.

Not adamantly opposed, but that solution feels like "chucking it all".

> Is it only the filters that go in there or the full path? I feel like it's hard to draw a line here

The problem really is the `in` filter, the only I've seen reported to grow that big(like a thousand values) for both GET and DELETE.

So perhaps we could do something like:

```http
POST /projects?status=eq.active
Content-Type: application/vnd.pgrst.big-in-filter+json

{
  "method": "DELETE"
, "in": [1,2,3,4,5,6]
}
```

That way we get to reuse path, headers and querystring.

**Comment by wolfgangwalther** [comment]
Hm. For such a specific solution that only targets a single filter, you can easily create an RPC - that should be good enough already.

**Comment by steve-chavez** [comment]
Yeah, but we kinda miss the "low-code" advantage with that. I thought we could have the above interface so client libraries can abstract the `in` filter problem.

**Comment by steve-chavez** [comment]
A [multipart type](https://developer.mozilla.org/en-US/docs/Web/HTTP/Basics_of_HTTP/MIME_types#multipart_types) could be an option to the [above](https://github.com/PostgREST/postgrest/issues/2125#issuecomment-1045214487).

**Comment by wolfgangwalther** [comment]
> The problem really is the `in` filter, the only I've seen reported to grow that big(like a thousand values) for both GET and DELETE.

I can see other use-cases, too: Passing a value in a JSON body should allow to work around some of the quoting/escaping issues/challenges we have in the query string.

> So perhaps we could do something like:
> 
> ```httpspec
> POST /projects?status=eq.active
> Content-Type: application/vnd.pgrst.big-in-filter+json
> 
> {
>   "method": "DELETE"
> , "in": [1,2,3,4,5,6]
> }
> ```

Only supporting that special cases just seems so... arbitrary.

> That way we get to reuse path, headers and querystring.

I can see how that's a positive. What about:

```httpspec
POST /projects?status=eq.active&id=in.refs->my_ids
Content-Type: application/vnd.pgrst.params-in-body+json

{
  "method": "DELETE"
, "body": "... original body ...",
, "refs": {
    "my_ids": [1,2,3,4,5,6]
  }
}
```

The idea would be to somehow allow **referencing** values in the body in the query string. This way they could be re-used for all kinds of filters, not only for `in` or so. And the general query structure is still entirely visible from the query string - including all filters.

Some ideas for syntax:
- variants for each operator, that would use a reference: `&id=in-ref.my_ids`
- a meta operator: `&id=in.ref.my_ids`
- a special character for access: `&id=in.$my_ids` or `&id=in.@my_ids`

Those would not be breaking changes, because they would only be interpreted if the special `Content-Type` is set. I think I like the last suggestion most...

**Comment by steve-chavez** [comment]
> The idea would be to somehow allow referencing values in the body in the query string.

Nice! I like that direction.

---

There's another concern I have. This type of bodies will definitely be big and parsing them in Haskell would lead us to have the same problem as https://github.com/PostgREST/postgrest/issues/690. So ideally we want to send the body directly to postgres and avoid processing it. With the `application/vnd.pgrst.params-in-body+json`, I believe we can work the query to somehow pass the `refs->my_ids` to the WHERE clause but this would not be possible for the `method` field since PostgREST would need it in advance..

> Perhaps some proxies don't even pass the body in a DELETE, we'd have to test that.
> I did a quick search for it, but only found other posts that said it was fine. Your link has a much more authoritative source, though...

I grow dissatisfied with answers with just authoritative/purist points of view and it has come to my attention that Elasticsearch [allows doing a GET with a request body](https://www.elastic.co/guide/en/elasticsearch/guide/current/_empty_search.html#get_vs_post).

How about if we support for GET/DELETE with a request body for these exceptional cases? Seems javascript `fetch` would be an issue(GET doesn't pass a request body) but I'm lead to believe that, most importantly, proxies would work fine.

References:

- https://evertpot.com/get-request-bodies/
- https://stackoverflow.com/questions/978061/http-get-with-request-body

**Comment by wolfgangwalther** [comment]
> There's another concern I have. This type of bodies will definitely be big and parsing them in Haskell would lead us to have the same problem as #690. So ideally we want to send the body directly to postgres and avoid processing it. With the `application/vnd.pgrst.params-in-body+json`, I believe we can work the query to somehow pass the `refs->my_ids` to the WHERE clause but this would not be possible for the `method` field since PostgREST would need it in advance..

Fully agree.

> I grow dissatisfied with answers with just authoritative/purist points of view and it has come to my attention that Elasticsearch [allows doing a GET with a request body](https://www.elastic.co/guide/en/elasticsearch/guide/current/_empty_search.html#get_vs_post).
> 
> How about if we support for GET/DELETE with a request body for these exceptional cases? Seems javascript `fetch` would be an issue(GET doesn't pass a request body) but I'm lead to believe that, most importantly, proxies would work fine.

Do we need to support `GET` with a request body at all? The idea was to use `QUERY` for that. As long as `DELETE` with a body works fine, we can avoid passing the method in the body.

I think I'd be fine with that.

---

For completeness, there is also a different way, which would be the most "REST-like", I guess:

When passing those parameters in the body, we break the "the same URI describes the same resource" promise - because with a different body, a completely different resource is targeted. This can be avoided by creating temporary resources in two http requests:
- First a `POST` request with the to-be-used filter in the body. Something like a temporary "params" resource on a special endpoint.
- Then run the main request and use the temporary resource (which should have a unique identifier - unique forever) as a filter. Something like an inner join or exists filter, basically.

The second request would then have the unique identifier in the URI and would not break that promise above anymore.

Those temporary resources could be represented as real database objects, e.g. temp tables or functions in the temporary schema. Although I don't have an idea, yet, how that would work with a connection pool.

**Comment by steve-chavez** [comment]
> Do we need to support GET with a request body at all? The idea was to use QUERY for that.

Another option could be the [REPORT](https://www.rfc-editor.org/rfc/rfc3253.html#page-25) method:

> A REPORT request is an extensible mechanism for obtaining information
   about a resource.  Unlike a resource property, which has a single
   value, the value of a report can depend on additional information
   specified in the REPORT request body and in the REPORT request
   headers.

According to this [blog post](https://evertpot.com/dropbox-post-api/#using-report-instead) it already works everywhere.

Other references:

- https://evertpot.com/webdav-features-for-http/#report--search
- https://www.iana.org/assignments/http-methods/http-methods.xhtml (REPORT is registered)
- https://httptoolkit.tech/blog/http-search-method/ (SEARCH is a previous version of QUERY)

**Comment by steve-chavez** [comment]
Branched off the DELETE with a body discussion to https://github.com/PostgREST/postgrest/issues/2314

**Comment by telewak** [comment]
any idea when this update could be implemented?

**Comment by telewak** [comment]
Thanks guys for the good work so far. Any idea when this feature could be implemented? i am waiting for it, passing arguments in request body when GETing or QUERYing for data

**Comment by steve-chavez** [comment]
DELETE with body will definitely come in v10. This one would be for the next version.

I think we should go with the SEARCH method as it's older and likely better supported. With that the [above proposal](https://github.com/PostgREST/postgrest/issues/2125#issuecomment-1046202010) would be:

```python
SEARCH /projects?status=eq.active&id=in.refs->my_ids
Content-Type: application/vnd.pgrst.params-in-body+json

{
 "refs": {
    "my_ids": [1,2,3,4,5,6]
  }
}
```

Or to avoid an special case in our filter operators we could:

```python
SEARCH /projects?status=eq.active&id=_in.my_ids
Content-Type: application/vnd.pgrst.params-in-body+json

{
  "my_ids": [1,2,3,4,5,6]
}
```

Where every operator with an underscore(`_op`) gets its value from a json key in the body.

**Comment by steve-chavez** [comment]
The SEARCH method could also allow us to pass parameters that are more bulky in nature, like for [PostGIS operators](https://github.com/PostgREST/postgrest/issues/223):

```python
SEARCH /projects?range_area=_st_within.area
Content-Type: application/vnd.pgrst.params-in-body+json

{
  "area": {
    "type": "Polygon",
    "coordinates": [
      [
        [
          -71.10034391283989,
          42.37385299961788
        ],
        [
          -71.10036939382553,
          42.373756895982865
        ],
        [
          -71.1002916097641,
          42.373745997623224
        ],
        [
          -71.1002641171217,
          42.37384408279195
        ],
        [
          -71.10034391283989,
          42.37385299961788
        ]
      ]
    ]
  }
}
```

**Comment by steve-chavez** [comment]
Right now we can only use [our operators](https://postgrest.org/en/stable/api.html#operators) against other values, we're missing operating against other identifiers.

Underscore operators could be used to operate on columns besides referencing values on the request body. This would allow doing [column to column comparison(asked before)](https://github.com/PostgREST/postgrest/issues/1105#issuecomment-395238192) without computed columns:

```python
SEARCH /PredictedGenes?Strand=eq.+&GeneStart=_gt.GeneEnd
```

Operating on a body value would still be allowed, `body` would be a reserved identifier:

```python
SEARCH /PredictedGenes?Strand=_eq.body->strand&GeneStart=_lt.GeneEnd

{ "Strand": "-"}
```

This would favor PATCH/DELETE as well.

---

> Why POST? We could invent a special body format, that would contain our filters. We'd use that for QUERY and DELETE by default - but would allow adding it to PATCH via content-type header. In the PATCH case the original payload would then be included in this body, too

@wolfgangwalther Just realized that If we implement the [set operator](https://github.com/PostgREST/postgrest/issues/465#issuecomment-1199858002) for PATCH, we wouldn't need a new `content-type` for it.

**Comment by steve-chavez** [comment]
Another option for underscore operators is to use a [jsonpath](https://www.postgresql.org/docs/12/datatype-json.html#TYPE-JSONPATH-VARIABLES) expression as the operand.

```http
DELETE /tbl?id=_eq.$[*].id

[
 {"id": 1},
 {"id": 2}
]
```

One good thing is that we'd get to reuse existing pg functionality. Another one that since `$` is invalid for the first char of an identifier, we can clearly separate json path values from identifiers(so column to column comparison would still be possible).

---

One difficulty I've noted with this is that we'd need the schema cache to cast the `json` to the column type.

```sql
select * from test.projects where id = (('{"a": 3}'::json)->>'a');
ERROR:  operator does not exist: integer = text

select * from test.projects where id = (('{"a": 3}'::json)->>'a')::int;
 id | name | client_id
----+------+-----------
  3 | IOS  |         2
(1 row)
```

It doesn't seem possible to obtain the json with an `unknown` casting.

**Comment by wolfgangwalther** [comment]
> Another one that since $ is invalid for the first char of an identifier, we can clearly separate json path values from identifiers(so column to column comparison would still be possible).

I don't really speak `jsonpath`, yet. Does an expression always have to start with `$`? I can't take that away immediately from skimming the docs on it.

**Comment by steve-chavez** [comment]
> Does an expression always have to start with $? 

Yes, otherwise there was a syntax error IIRC. You can also use `$var` for functions that pass variables.

I no longer think that json path is safe to expose though, made a comment on https://github.com/PostgREST/postgrest/issues/1883#issuecomment-1204761390

**Comment by chelodegli** [comment]
I'd love to see this feature.
I have found aditional information for the new Query method, maybe it's somehow useful:
[https://datatracker.ietf.org/doc/draft-ietf-httpbis-safe-method-w-body/](url)

**Comment by steve-chavez** [comment]
Elaborating on the [above](https://github.com/PostgREST/postgrest/issues/2125#issuecomment-1200054135). Instead of underscore operators, I think dollar operators would be better: 

```http
SEARCH /PredictedGenes?Strand=$eq.$body.strand&GeneStart=$lt.GeneEnd
{ "Strand": "-"}
```

It looks more differentiable/clearer that it's our own syntax. And instead of having `body` as a reserved identifier, we can have namespaces(that can be used in other cases) like `$body.col1` and also support a shortened version like `$b.col1`.

**Comment by steve-chavez** [comment]
What if we use the our query grammar in the media type parameters. Like:

```http
SEARCH /PredictedGenes
Content-Type: application/vnd.pgrst.search+json; Strand.eq=$Strand; GeneStart.lt=GeneEnd

{ "Strand": "-"}
```

This would make it easier to implement as the query string grammar is not modified.

Media type parameters are also proposed for PATCH https://github.com/PostgREST/postgrest/issues/465#issuecomment-1566188612.

---

Also, maybe later this could be used for solving client-side transactions(https://github.com/PostgREST/postgrest/issues/286).

```python
POST / HTTP/1.1
Content-Type: multipart/form-data;boundary="boundary"

--boundary
Content-Type: application/vnd.pgrst.search+json; col.eq=$field;
Content-Disposition: form-data; name="tbl";

{..}

--boundary
Content-Type: application/vnd.pgrst.patch+json; another_col.eq=$another_field;
Content-Disposition: form-data; name="other_tbl";

{...}
```

**Comment by steve-chavez** [comment]
Wondering if we should do this. This recent article:

https://bessey.dev/blog/2024/05/24/why-im-over-graphql/#query-parsing

Mentions how sending a long list over the POST body can be abused to cause OOM, and how mitigating this is a pain in GraphQL.

So for us, since we use the URL for our query language, the URL size limit has always served as a built-in defense mechanism.

In a way, this limitation has always been a feature. Perhaps we should document it as such? And try to give a better error message if it happens?
